File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_board_game.java
import java.util.*;
import java.io.*;
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	static int dis[]; static boolean vis[];
	public static void main(String[] args) throws IOException{	
	    int st = readInt(), ed = readInt(), maxV = (int)3e5;
	    dis = new int[maxV+1]; vis = new boolean[maxV+1];
	    Queue<Integer> q = new LinkedList<>();
	    q.add(st); vis[st]=true;
	    while(!q.isEmpty()){
	        int cur = q.poll();
	        if(cur -1 >= 0 && !vis[cur-1]) push(q, cur, cur-1);
	        if(cur -3 >= 0 && !vis[cur-3]) push(q, cur, cur-3);
	        if(3*cur < vis.length && !vis[3*cur]) push(q, cur, 3*cur);
	        if(cur % 2 ==0 && !vis[cur/2]) push(q, cur, cur/2);
	        if(vis[ed]) { System.out.println(dis[ed]); break;}
	    }
	}
	static void push(Queue<Integer> q, int cur, int next) {
		q.add(next); vis[next]=true; dis[next]=dis[cur]+1;
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_board_game.cpp
#include <bits/stdc++.h>
using namespace std;
const int MV = 300003;
int st, ed, dis[MV]; queue<int> q; bool vis[MV];
int main(){
    cin >> st >> ed;
    q.push(st); vis[st]=1;
    while(!q.empty()){
        int cur = q.front(); q.pop();
        if(cur == ed) { cout << dis[ed] << endl; return 0; }
        if(cur -1 >= 0 && !vis[cur-1]){
            q.push(cur-1); dis[cur-1] = dis[cur] + 1; vis[cur-1]=1;
        }
        if(cur -3 >= 0 && !vis[cur-3]){
            q.push(cur-3); dis[cur-3] = dis[cur] + 1; vis[cur-3]=1;
        }
        if(3*cur < MV && !vis[3*cur]){
            q.push(3*cur); dis[3*cur] = dis[cur] + 1; vis[3*cur]=1;
        }
        if(cur % 2 ==0 && !vis[cur/2]){
            q.push(cur/2); dis[cur/2] = dis[cur] + 1; vis[cur/2]=1;
        }
    }
}